bentoml/_internal/types.py

```diff
@@ -43,6 +43,130 @@
 JSONSerializable = t.NewType("JSONSerializable", object)
 
 
+T = t.TypeVar("T")
+
+
+class TypeRef(t.Generic[T]):
+    """
+    TypeRef provides solutions for several conflicts when applying lazy dependencies,
+        type annotations and runtime class checking.
+    It works both for runtime and type checking phases.
+
+    * conflicts 1
+
+    isinstance(obj, class) requires importing the class first, which breaks
+    lazy dependencies
+
+    solution:
+    >>> TypeRef("numpy.ndarray").isinstance(obj)
+
+    * conflicts 2
+
+    `isinstance(obj, str)` will narrow obj types down to str. But it only works for the
+    case that the class is the type at the same time. For numpy.ndarray which the type
+    is actually numpy.typing.NDArray, we had to hack the type checking.
+
+    solution:
+    >>> if TYPE_CHECKING:
+    >>>     from numpy.typing import NDArray
+    >>> TypeRef["NDArray"]("numpy.ndarray").isinstance(obj)`
+    >>> #  this will narrow the obj to NDArray with PEP-647
+
+    * conflicts 3
+
+    compare/refer/map classes before importing them.
+
+    >>> HANDLER_MAP = {
+    >>>     TypeRef("numpy.ndarray"): ndarray_handler,
+    >>>     TypeRef("pandas.DataFrame"): pdframe_handler,
+    >>> }
+    >>>
+    >>> HANDLER_MAP[TypeRef(numpy.ndarray)]](array)
+    >>> TypeRef("numpy.ndarray") == numpy.ndarray
+    """
+
+    @t.overload
+    def __init__(self, module_or_cls: str, qualname: str) -> None:
+        """TypeRef("numpy", "ndarray")"""
+
+    @t.overload
```

Comment:
I think this will throw the pylint error
```suggestion
    @overload
```

Reply:
error happens after your patch
```
bentoml/_internal/types.py:92:6: F821 undefined name 'overload'
bentoml/_internal/types.py:97:5: F811 redefinition of unused '__init__' from line 93
```